autorest/azure/devicetoken.go

```diff
@@ -0,0 +1,196 @@
+package azure
+
+/*
+  This file is largely based on rjw57/oauth2device's code, with the follow differences:
+   * scope -> resource, and only allow a single one
+   * receive "Message" in the DeviceCode struct and show it to users as the prompt
+   * azure-xplat-cli has the following behavior that this emulates:
+     - does not send client_secret during the token exchange
+     - sends resource again in the token exchange request
+*/
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest"
+)
+
+const (
+	// OAuthDeviceEndpoint is Azure's OAuth2 Device Flow Endpoint
+	OAuthDeviceEndpoint = "https://login.microsoftonline.com/common/oauth2/devicecode"
+	// OAuthTokenEndpoint is Azure's OAuth2 Token Endpoint
+	OAuthTokenEndpoint = "https://login.microsoftonline.com/common/oauth2/token"
+
+	// authAPIVersionQueryParamName is the name
+	authAPIVersionQueryParamName  = "api-version"
+	authAPIVersionQueryParamValue = "1.0"
+)
+
+var (
+	// ErrDeviceGeneric represents an unknown error from the token endpoint when using device flow
+	ErrDeviceGeneric = fmt.Errorf("Error while retrieving OAuth token: Unknown Error")
+
+	// ErrDeviceAccessDenied represents an access denied error from the token endpoint when using device flow
+	ErrDeviceAccessDenied = fmt.Errorf("Error while retrieving OAuth token: Access Denied")
+
+	// ErrDeviceAuthorizationPending represents the server waiting on the user to complete the device flow
+	ErrDeviceAuthorizationPending = fmt.Errorf("Error while retrieving OAuth token: Authorization Pending")
+
+	// ErrDeviceCodeExpired represents the server timing out and expiring the code during device flow
+	ErrDeviceCodeExpired = fmt.Errorf("Error while retrieving OAuth token: Code Expired")
+
+	// ErrDeviceSlowDown represents the service telling us we're polling too often during device flow
+	ErrDeviceSlowDown = fmt.Errorf("Error while retrieving OAuth token: Slow Down")
+)
```

Comment:
Yes. If the user does not wish to use the automatic looping of WaitForUserCompletion then they need these errors to compare against to know how to proceed.